sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -939,68 +954,65 @@ class GpuDynamicPartitionDataConcurrentWriter(
+      statsTrackers.foreach(_.newBatch(status.writerStatus.outputWriter.path(), batch))
+      status.writerStatus.recordsInFile += batch.numRows()
+      status.writerStatus.outputWriter.writeAndClose(batch, statsTrackers)
+    } else {
+      val (tabs, dataTypes) = withResource(batch) { batch =>
+        withResource(GpuColumnVector.from(batch)) { table =>
+          val splits =
+            splitToFitMaxRecords(
+              table,
               maxRecordsPerFile,
-              status.writerStatus.recordsInFile,
-              table.getRowCount()
-            )
-
-            val dataTypes = GpuColumnVector.extractTypes(batch)
-
-            var needNewWriter = status.writerStatus.recordsInFile >= maxRecordsPerFile
-            withResource(table.contiguousSplit(splitIndexes: _*)) {tabs =>
-              tabs.foreach(b => {
-                if (needNewWriter) {
-                  status.writerStatus.fileCounter += 1
-                  assert(status.writerStatus.fileCounter <= MAX_FILE_COUNTER,
-                    s"File counter ${status.writerStatus.fileCounter} " +
-                      s"is beyond max value $MAX_FILE_COUNTER")
-                  status.writerStatus.outputWriter.close()
-
-                  // start a new writer
-                  val w = newWriter(partitionDir, None, status.writerStatus.fileCounter)
-                  status.writerStatus.outputWriter = w
-                  status.writerStatus.recordsInFile = 0L
-                }
-                val cb = withResource(b.getTable()) {tab =>
-                  GpuColumnVector.from(tab, dataTypes)
-                }
-                statsTrackers.foreach(_.newBatch(status.writerStatus.outputWriter.path(), cb))
-                status.writerStatus.recordsInFile += b.getRowCount()
-                status.writerStatus.outputWriter.writeAndClose(cb, statsTrackers)
-                needNewWriter = true
-              })
-            }
-          }
+              status.writerStatus.recordsInFile)
+          (splits, GpuColumnVector.extractTypes(batch))
         }
       }
-      status.tableCaches.clear()
-      status.deviceBytes = 0
+      var needNewWriter = status.writerStatus.recordsInFile >= maxRecordsPerFile
+      withResource(tabs) { _ =>
+        tabs.foreach(b => {
+          if (needNewWriter) {
+            status.writerStatus.fileCounter += 1
+            assert(status.writerStatus.fileCounter <= MAX_FILE_COUNTER,
+              s"File counter ${status.writerStatus.fileCounter} " +
+                  s"is beyond max value $MAX_FILE_COUNTER")
+            status.writerStatus.outputWriter.close()
+
+            // start a new writer
+            val w = newWriter(partitionDir, None, status.writerStatus.fileCounter)
+            status.writerStatus.outputWriter = w
+            status.writerStatus.recordsInFile = 0L
+          }
+          val cb = GpuColumnVector.from(b.getTable, dataTypes)
```

Comment:
Shouldn't we close `b` here and null out the corresponding `tabs` entry so we can close these early?  Looks like everything except the last entry in `tabs` is held on longer than it needs to be.